core/shared/src/main/scala/scalaz/zio/stream/Stream.scala

```diff
@@ -0,0 +1,762 @@
+            .flatMap(step => sink.extract(Sink.Step.state(step)))
+      )
+
+  /**
+   * Statefully maps over the elements of this stream to produce new elements.
+   */
+  def scan[S1, B](s1: S1)(f1: (S1, A) => (S1, B)): Stream[E, B] = new Stream[E, B] {
+    override def foldLazy[E1 >: E, B1 >: B, S](s: S)(cont: S => Boolean)(f: (S, B1) => IO[E1, S]): IO[E1, S] =
+      self
+        .foldLazy[E1, A, (S, S1)](s -> s1)(tp => cont(tp._1)) {
+          case ((s, s1), a) =>
+            val (s2, b) = f1(s1, a)
+
+            f(s, b).map(s => s -> s2)
+        }
+        .map(_._1)
+  }
+
+  /**
+   * Statefully and effectfully maps over the elements of this stream to produce
+   * new elements.
+   */
+  final def scanM[E1 >: E, S1, B](s1: S1)(f1: (S1, A) => IO[E1, (S1, B)]): Stream[E1, B] = new Stream[E1, B] {
+    override def foldLazy[E2 >: E1, B1 >: B, S](s: S)(cont: S => Boolean)(f: (S, B1) => IO[E2, S]): IO[E2, S] =
+      self
+        .foldLazy[E2, A, (S, S1)](s -> s1)(tp => cont(tp._1)) {
+          case ((s, s1), a) =>
+            f1(s1, a).flatMap {
+              case (s1, b) =>
+                f(s, b).map(s => s -> s1)
+            }
+        }
+        .map(_._1)
+  }
+
+  /**
+   * Takes the specified number of elements from this stream.
+   */
+  final def take(n: Int): Stream[E, A] =
+    self.zipWithIndex.takeWhile(_._2 < n).map(_._1)
+
+  /**
+   * Takes all elements of the stream for as long as the specified predicate
+   * evaluates to `true`.
+   */
+  def takeWhile(pred: A => Boolean): Stream[E, A] = new Stream[E, A] {
+    override def fold[E1 >: E, A1 >: A, S](s: S)(f: (S, A1) => IO[E1, Step[S]]): IO[E1, Step[S]] =
+      self.fold[E1, A, S](s)((s, a) => if (pred(a)) f(s, a) else IO.now(Step.Stop(s)))
+
+    override def foldLazy[E1 >: E, A1 >: A, S](s: S)(cont: S => Boolean)(f: (S, A1) => IO[E1, S]): IO[E1, S] =
+      self
+        .foldLazy[E1, A, (Boolean, S)](true -> s)(tp => tp._1 && cont(tp._2)) {
+          case ((_, s), a) =>
+            if (pred(a)) f(s, a).map(true -> _)
+            else IO.now(false             -> s)
+        }
+        .map(_._2)
+  }
+
+  /**
+   * Converts the stream to a managed queue. After managed queue is used, the
+   * queue will never again produce values and should be discarded.
+   */
+  final def toQueue[E1 >: E, A1 >: A](capacity: Int = 1): Managed[Nothing, Queue[Take[E1, A1]]] =
+    for {
+      queue    <- Managed.liftIO(Queue.bounded[Take[E1, A1]](capacity))
+      offerVal = (a: A) => queue.offer(Take.Value(a)).void
+      offerErr = (e: E) => queue.offer(Take.Fail(e))
+      enqueuer = (self.foreach[E](offerVal).catchAll(offerErr) *> queue.offer(Take.End).forever).fork
+      _        <- Managed(enqueuer)(_.interrupt)
+    } yield queue
+
+  /**
+   * Applies a transducer to the stream, which converts one or more elements
+   * of type `A` into elements of type `C`.
+   */
+  final def transduce[E1 >: E, A1 >: A, C](sink: Sink[E1, A1, A1, C]): Stream[E1, C] =
```

Comment:
@jdegoes Could you review this implementation of transduce? I had to add the last `extract` attempt to get the test passing.